plugins/kubernetes/src/components/KubernetesContent/KubernetesContent.tsx

```diff
@@ -105,15 +113,63 @@ export const KubernetesContent = ({ entity }: KubernetesContentProps) => {
   >(undefined);
   const [error, setError] = useState<string | undefined>(undefined);
 
+  const configApi = useApi(configApiRef);
+  const clusters: Config[] = configApi.getConfigArray('kubernetes.clusters');
+  const allAuthProviders: string[] = clusters.map(c =>
+    c.getString('authProvider'),
+  );
+  const authProviders: string[] = [...new Set(allAuthProviders)];
+
+  // Create mapping for supported auth providers to their API refs
+  const authProviderApiMapping: Map<string, ApiRef<OAuthApi>> = new Map<
+    string,
+    ApiRef<OAuthApi>
+  >();
+  authProviderApiMapping.set('google', googleAuthApiRef);
+
+  // These and their authProviders have to be instantiated outside of the useEffect hook
+  const kubernetesAuthProviderMap: Map<
+    string,
+    KubernetesAuthProvider
+  > = new Map<string, KubernetesAuthProvider>();
+  // TODO: Query Kubernetes backend to get info on which auth providers are in use
+  // For each auth type, create instance of KubernetesAuthProvider
+  for (const authProviderStr of authProviders) {
+    let api;
+    if (authProviderApiMapping.has(authProviderStr)) {
+      api = useApi(authProviderApiMapping.get(authProviderStr)!);
```

Comment:
This causes this linting error to be thrown: `React Hook "useApi" may be executed more than once. Possibly because it is called in a loop. React Hooks must be called in the exact same order in every component render  react-hooks/rules-of-hooks`. I'm not sure how to deal with this in a way that doesn't involve loading in the same set of providers/APIs (ex. all the ones that we support for Kubernetes auth) each time/in a hardcoded manner.

It seems that any dynamic usage of useApi (ex. in a way that only involves loading in the K8s auth related APIs that are defined in the app-config) would result in this error being thrown/noted, due to inconsistent useApi invocation order.

Reply:
Yeah that's one of the quirks of hooks in React in general because of how they are implemented - they have to always be invoked in the same order. That's why you'll almost always find them at the very top of component code, and never in a loop or conditional.

Reply:
So I haven't looked at the rest of the PR yet, but it seems that you could store the provider key-to-instance map in a provider of sorts, that you can fetch once at the top as `useKubernetesAuthProviders()`.

If that sounds like circular reasoning (how would that one then be implemented), you could likely instantiate it at plugin boot time, something along the lines of

```ts
export const plugin = createPlugin({
  id: 'myKubernetesPlugin',
  apis: [
    createApiFactory({
      api: kubernetesAuthProvidersRef,
      deps: { configApi: configApiRef, googleApi: googleApiRef },
      factory({ configApi, googleApi }) {
        // ... create the various supported providers
        return new KubernetesAuthProviders(...);
      },
    }),
  ],
});
```

Reply:
Gotcha, instantiation at plugin creation time makes sense. Would I be able to, in this section of logic (presumably in here: https://github.com/spotify/backstage/blob/2d73c037a0affd3b1c2eba6878543a37cd53ba1f/plugins/kubernetes/src/plugin.ts) load in the configApi outside of this `createPlugin` invocation, use it to read the `app-config` and get the list of different `authProvider`s that are defined across the various `kubernetes.clusters` in the config, and then create an `kubernetesAuthProvidersRef` that uses the various auth APIs correlated with the defined `authProvider`s ? 

Seems like I would be able to do this as this doesn't use `useApi`? And I would simply be able to do one `useApi(kubernetesAuthProvidersRef)` in the `KubernetesContent.tsx`. I'll give this a try later.

Reply:
@RochesterinNYC Did actually run in to this with the auth providers and thought about some kinda meta-factory that can read config and generate factories. Didn't go down that path though, because I don't see any big win with it and it adds a lot of complexity.

I think the sane way to go for now is to list all the auth APIs that you have implementations for, and then allow more granular access and customization by overriding the default factory in the app.

A possible general solution I think could be the ability to optionally depend on APIs, as in if they're available in the app, great, use it, but if it isn't , we don't care. That or lazy loading, or both. That's something we could look at bringing in if we feel there's a need, both for the `useApi` and `createApiFactory` routes.

Either way you need to enable the possibility for the consumer to provide their own APIs in some way, because an organization that uses Backstage may have their completely custom internal authentication provider, implemented with a custom and unknown to us `ApiRef`.

Reply:
I amended the top commit to basically create a `KubernetesAuthProviders` "API" as @freben suggested that stores/holds instances of the `KubernetesAuthProvider`s that are currently supported and is created at plugin time instead of `KubernetesComponent` render/creation time.

The google auth workflow still works as expected.

Ready for review again @mclarke47 @Rugvip. I think we should at some point seek to merge this in as it's currently working for Google auth and then make more PRs/improvements and etc. instead of aiming to make this one massive, all-encompassing PR that fits exactly how we want to do Kubernetes auth abstractions/interface/UX wise overall.

Reply:
Yeah I'm up for shipping and iterating too